### ISSUE TITLE
RiakCS does not have documentation covering nodejs clients connecting over http [JIRA: RCS-260]

### ISSUE THREAD

**Issue by StuAtGit** [basho/riak_cs#1211]
It looks like RiakCS does not support the aws s3 interface.

I'm attempting to the aws nodejs client like so:

```
var awsConfig = {
    accessKeyId:riakCSUsername,
    secretAccessKey:riakCSPassword,
    sslEnabled:false,
    httpOptions:{
        proxy:"http://" + riakCSServer + ":" + riakCSPort
    },
    signatureVersion: 'v2'
};

var s3client = new aws.S3( new aws.Config(awsConfig) );
s3client.getObject( {
        Bucket: bucket,
        Key: key
    }, function( result, err ) {
        console.log("S3 result: " + result + " error: " + err );
    }
);
```

And I get:
S3 result: TypeError: Cannot set property 'Timestamp' of undefined error: null

This seems to be along the lines of:
http://lists.basho.com/pipermail/riak-users_lists.basho.com/2015-February/016846.html

And there does seem to be some workaround, but all that is stated is there is a 'simple fix'.

I've been digging around for a while, and noted several issues closed with the reasoning along the lines of "we follow the HTTP spec, and amazon does it wrong"... so I'm wondering if it's a reasonable expectation to think that riak cs is compatible with s3 clients?

I'm a little new to the best practices here - I'm in this spot because my company runs several large riak clusters, and I was told "its just like s3", "it's like a distributed filesystem" (not that I can see??, but OK), and I have some old code using a python library from aws that I was told works. So I thought the aws sdk for nodejs might work too..... is this reasonable?

I tried the basho nodejs client as well, but:

1) I spent over an hour reading the documentation and going through the code just to figure out how to sent authorization credentials..

2) The code I try with the basho client never calls the ping callback - so I have no idea what's going wrong.

**Comment by StuAtGit** [comment]
Note that when I tweak the code for my own personal aws repo, it works as expected:

```
var awsConfig = {
    accessKeyId:riakCSUsername,
    secretAccessKey:riakCSPassword,
    sslEnabled:true
};

var s3client = new aws.S3( new aws.Config(awsConfig) );
s3client.getObject( {
        Bucket: bucket,
        Key: key
    }, function( result, err ) {
        console.log("S3 result: " + result + " error: " + err );
    }
);
```

**Comment by StuAtGit** [comment]
Here is the basho-riak-client code:

```
var riakClient = new riak.Client(
    new riak.Cluster({
            nodes: [
                new riak.Node({
                    remoteAddress: riakCSServer,
                    remotePort: riakCSPort,
                    auth: {
                        user: riakCSUsername,
                        password: riakCSPassword
                    },
                    cork: false
                })
            ]
        }
    )
);

riakClient.ping( function(err, result) {
    if( err ) {
        console.log("RiakCS Error: " + err );
    } else {
        console.log("RiakCS success!: " + result);
    }
});

riakClient.fetchValue( {
    bucket:bucket,
    key:key
}, function( err, result ) {
    console.log("riak client said: " + result + " " + err);
});
```

This prints out nothing at all.

**Comment by StuAtGit** [comment]
I'm not sure if the basho client is using protocol buffers - I saw some notes saying that it might. If so I may not have access to that port - is there a way to ensure the basho node client uses http?

I just need something that works within nodejs. I don't really care if it's s3 or not - just need to get and post some data, is all. Riak worked fine, but RiakCS just won't cooperate with any of the existing javascript clients I've tried.

**Comment by StuAtGit** [comment]
I actually played around with pointing the aws-sdk at aws s3, with it set to v2, that is, point it at my personal aws account, and do

```
var awsConfig = {
    accessKeyId:riakCSUsername,
    secretAccessKey:riakCSPassword,
    sslEnabled:true
};
```

And got the same results as against riakcs.
The reply I got from the aws-sdk people:

"AWS.Signers.V2 is not supported by Amazon S3. S3 supports a modified version of Signature Version 2. In the SDK this is implemented by AWS.Signers.S3. You shouldn't have to specify this signer as the AWS.S3 client picks this up by default. "

https://github.com/aws/aws-sdk-js/issues/678

Soo... not sure where to go next with this. Since I believe the basho client only supports protobufs, and the nodejs aws-sdk doesn't work with riakcs, I guess I'll just spin up a microservice in another language?

**Comment by StuAtGit** [comment]
Ok.. some progress.
After some debugging through the s3 client with the "S3" signature version, it looks like that is was riakcs supports.

And, the httpOptions "proxy" is not what you use to change the endpoint, so I need to fix that as well.
So, I've gotten to here:

```
var awsEndpoint = new aws.Endpoint(riakCSServer);
awsEndpoint.port = riakCSPort;
awsEndpoint.protocol = "http";
awsEndpoint.href = "http://" + riakServer + ":" + riakCSPort + "/";

var awsConfig = {
    accessKeyId:riakCSUsername,
    secretAccessKey:riakCSPassword,
    sslEnabled:false,
    endpoint:awsEndpoint,
    signatureVersion: 's3'
};

var s3client = new aws.S3( new aws.Config(awsConfig) );
s3client.getObject( {
        Bucket: bucket,
        Key: key
    }, function( err, result ) {
        console.log("S3 result: " + result + " error: " + err );
    }
);
```

==> S3 result: null error: NoSuchBucket: The specified bucket does not exist

I think the bucket should exist, but that is likely a problem on my end - will update once I know for sure.

So it looks like the issue is not a bug in the code per-se, but certainly there are some bugs in the documentation on both sides 
- aws doesn't do a good job of documenting how to point at other clouds (which is sort-of understandable, given the business model).
- riak basho client documentation doesn't do a good job of explaining how to use it with anything but very trivial scenarios or explaining/clarifying whether/how the client is using protobuf vs http.
- there is no basho documentation that I found on how to use js aws clients with riakcs. I would think it would be in basho interests to improve this? But I don't know.. maybe not?

**Comment by StuAtGit** [comment]
Confirmed that it works now.

If someone can point me to some basho documentation that includes a few examples along the lines of:

### How to connect to RiakCS using http and aws-sdk in Node:

```
var awsEndpoint = new aws.Endpoint(riakCSServer);
awsEndpoint.port = riakCSPort;
awsEndpoint.protocol = "http";
awsEndpoint.href = "http://" + riakServer + ":" + riakCSPort + "/";

var awsConfig = {
    accessKeyId:riakCSUsername,
    secretAccessKey:riakCSPassword,
    sslEnabled:false,
    endpoint:awsEndpoint,
    signatureVersion: 's3'
};

var s3client = new aws.S3( new aws.Config(awsConfig) );
s3client.getObject( {
        Bucket: bucket,
        Key: key
    }, function( err, result ) {
        console.log("S3 result: " + result + " error: " + err );
    }
);
```

### How to connect to riakcs using the basho node client over http

```
?
```

(I don't know how to do this)

I'd consider this issue closed.

**Comment by kuenishi** [comment]
I would like you to make sure that Riak CS provides S3-like API and Riak provides original protobuf API. You look like you've set up Riak CS cluster and trying to use S3 API. That is why [riak-nodejs-client](https://github.com/basho/riak-nodejs-client) does not work with S3 API. Moreover, riak-nodejs-client is only for protobuf IIRC. They are different interfaces.

We do not provide any official client libraries for Riak CS and let people choose what to use.

**Comment by StuAtGit** [comment]
Yes, I did figure that out, as my report above outlines. That is why I changed the name of the bug to cover the lack of documentation of the s3 client. Since s3 protocol is defined by the amazon sdk and documentation, it would make sense to document how to connect with riakcs with what is the reference implementation of the s3 protocol, in a few common languages, using the reference sdk for the protocol (i.e Amazon's).

Investing so much effort into the server-side, and then refusing to provide (or at least document how to use the system with) client libraries seems a bit odd.

At the very least the documentation could be clearer about the fact that the riak clients do not work with the riak backend when using the s3 protocol.
